pkg/automationconfig/automation_config_builder.go

```diff
@@ -161,6 +162,30 @@ func (b *Builder) Build() (AutomationConfig, error) {
 		hostnames[i] = fmt.Sprintf("%s-%d.%s", b.name, i, b.domain)
 	}
 
+	// If we are upgrading, we can't increase featureCompatibilityVersion
+	// as that will make the agent never reach goal state
+	if len(b.previousAC.Processes) > 0 && b.fcv == "" {
+
+		// Create a x.y.0 version from FCV x.y
+		previousFCV := b.previousAC.Processes[0].FeatureCompatibilityVersion
+		previousFCVsemver, err := semver.Make(fmt.Sprintf("%s.0", previousFCV))
+		if err != nil {
+			return AutomationConfig{}, errors.Errorf("can't compute semver version from previous FeatureCompatibilityVersion %s", previousFCV)
+		}
+
+		currentVersionSemver, err := semver.Make(b.mongodbVersion)
+		if err != nil {
+			return AutomationConfig{}, errors.Errorf("current MongoDB version is not a valid semver version: %s", b.mongodbVersion)
+		}
+
+		// We would increase FCV here.
+		// Note: in theory this will also catch upgrade like 4.2.0 -> 4.2.1 but we don't care about those
+		// as they would not change the FCV
+		if currentVersionSemver.GT(previousFCVsemver) {
+			b.fcv = previousFCV
+		}
+	}
```

Comment:
[nit] this is a lot of logic to also include in the Build method, maybe we should extract this out into a separate function.